File: app/src/main/java/com/quran/labs/androidquran/ui/AudioManagerActivity.java
package com.quran.labs.androidquran.ui;

import android.content.Intent;
import android.content.IntentFilter;
import android.os.Bundle;
import android.support.v4.content.LocalBroadcastManager;
import android.support.v7.app.ActionBar;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.quran.labs.androidquran.QuranApplication;
import com.quran.labs.androidquran.R;
import com.quran.labs.androidquran.common.QariItem;
import com.quran.labs.androidquran.data.SuraAyah;
import com.quran.labs.androidquran.service.QuranDownloadService;
import com.quran.labs.androidquran.service.util.DefaultDownloadReceiver;
import com.quran.labs.androidquran.service.util.QuranDownloadNotifier;
import com.quran.labs.androidquran.service.util.ServiceIntentHelper;
import com.quran.labs.androidquran.util.AudioManagerUtils;
import com.quran.labs.androidquran.util.AudioUtils;
import com.quran.labs.androidquran.util.QariDownloadInfo;
import com.quran.labs.androidquran.util.QuranFileUtils;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.observers.DisposableSingleObserver;


public class AudioManagerActivity extends QuranActionBarActivity
    implements DefaultDownloadReceiver.SimpleDownloadListener {
  private static final String AUDIO_DOWNLOAD_KEY = "AudioManager.DownloadKey";

  private ProgressBar progressBar;
  private Disposable disposable;
  private ShuyookhAdapter shuyookhAdapter;
  private RecyclerView recyclerView;
  private DefaultDownloadReceiver downloadReceiver;
  private String basePath;
  private List<QariItem> qariItems;

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    QuranApplication quranApp = (QuranApplication) getApplication();
    quranApp.refreshLocale(this, false);

    super.onCreate(savedInstanceState);
    final ActionBar ab = getSupportActionBar();
    if (ab != null) {
      ab.setTitle(R.string.audio_manager);
      ab.setDisplayHomeAsUpEnabled(true);
    }

    setContentView(R.layout.audio_manager);

    recyclerView = (RecyclerView) findViewById(R.id.recycler_view);
    recyclerView.setHasFixedSize(true);
    recyclerView.setLayoutManager(new LinearLayoutManager(this));
    recyclerView.setItemAnimator(new DefaultItemAnimator());

    qariItems = AudioUtils.getQariList(this);
    shuyookhAdapter = new ShuyookhAdapter(qariItems);
    recyclerView.setAdapter(shuyookhAdapter);

    progressBar = (ProgressBar) findViewById(R.id.progress);

    basePath = QuranFileUtils.getQuranAudioDirectory(this);
    getShuyookhData();
  }

  private void getShuyookhData() {
    if (disposable != null) {
      disposable.dispose();
    }
    disposable = AudioManagerUtils.shuyookhDownloadObservable(basePath, qariItems)
        .observeOn(AndroidSchedulers.mainThread())
        .subscribeWith(mOnDownloadInfo);
  }

  @Override
  protected void onResume() {
    super.onResume();
    downloadReceiver = new DefaultDownloadReceiver(this,
      QuranDownloadService.DOWNLOAD_TYPE_AUDIO);
    downloadReceiver.setCanCancelDownload(true);
    LocalBroadcastManager.getInstance(this).registerReceiver(downloadReceiver,
        new IntentFilter(QuranDownloadNotifier.ProgressIntent.INTENT_NAME));
    downloadReceiver.setListener(this);
  }

  @Override
  protected void onPause() {
    downloadReceiver.setListener(null);
    LocalBroadcastManager.getInstance(this).unregisterReceiver(downloadReceiver);
    super.onPause();
  }

  @Override
  protected void onDestroy() {
    disposable.dispose();
    super.onDestroy();
  }

  private DisposableSingleObserver<List<QariDownloadInfo>> mOnDownloadInfo =
      new DisposableSingleObserver<List<QariDownloadInfo>>() {
        @Override
        public void onSuccess(List<QariDownloadInfo> downloadInfo) {
          progressBar.setVisibility(View.GONE);
          shuyookhAdapter.setDownloadInfo(downloadInfo);
          shuyookhAdapter.notifyDataSetChanged();
        }

        @Override
        public void onError(Throwable e) {
        }
      };

  private View.OnClickListener mOnClickListener = new View.OnClickListener() {
    @Override
    public void onClick(View v) {
      int position = recyclerView.getChildAdapterPosition(v);
      if (position != RecyclerView.NO_POSITION) {
        QariDownloadInfo info = shuyookhAdapter.getSheikhInfoForPosition(position);
        if (info.downloadedSuras.size() != 114) {
          download(qariItems.get(position));
        }
      }
    }
  };

  private void download(QariItem qariItem) {
    String baseUri = basePath + qariItem.getPath();
    boolean isGapless = qariItem.isGapless();

    String sheikhName = qariItem.getName();
    Intent intent = ServiceIntentHelper.getDownloadIntent(this,
        AudioUtils.getQariUrl(qariItem),
        baseUri, sheikhName, AUDIO_DOWNLOAD_KEY, QuranDownloadService.DOWNLOAD_TYPE_AUDIO);
    intent.putExtra(QuranDownloadService.EXTRA_START_VERSE, new SuraAyah(1, 1));
    intent.putExtra(QuranDownloadService.EXTRA_END_VERSE, new SuraAyah(114, 6));
    intent.putExtra(QuranDownloadService.EXTRA_IS_GAPLESS, isGapless);
    startService(intent);

    AudioManagerUtils.clearCacheKeyForSheikh(qariItem);
  }

  @Override
  public boolean onOptionsItemSelected(MenuItem item) {
    if (item.getItemId() == android.R.id.home) {
      finish();
      return true;
    }
    return super.onOptionsItemSelected(item);
  }

  @Override
  public void handleDownloadSuccess() {
    getShuyookhData();
  }

  @Override
  public void handleDownloadFailure(int errId) {
    getShuyookhData();
  }

  private class ShuyookhAdapter extends RecyclerView.Adapter<SheikhViewHolder> {
    private final LayoutInflater mInflater;
    private final List<QariItem> mQariItems;
    private final Map<QariItem, QariDownloadInfo> mDownloadInfoMap;

    ShuyookhAdapter(List<QariItem> items) {
      mQariItems = items;
      mDownloadInfoMap = new HashMap<>();
      mInflater = LayoutInflater.from(AudioManagerActivity.this);
    }

    void setDownloadInfo(List<QariDownloadInfo> downloadInfo) {
      for (QariDownloadInfo info : downloadInfo) {
        mDownloadInfoMap.put(info.qariItem, info);
      }
    }

    @Override
    public SheikhViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
      return new SheikhViewHolder(mInflater.inflate(R.layout.audio_manager_row, parent, false));
    }

    @Override
    public void onBindViewHolder(SheikhViewHolder holder, int position) {
      holder.name.setText(mQariItems.get(position).getName());

      QariDownloadInfo info = getSheikhInfoForPosition(position);
      int fullyDownloaded = info.downloadedSuras.size();
      holder.quantity.setText(
          getResources().getQuantityString(R.plurals.files_downloaded,
            fullyDownloaded, fullyDownloaded));
    }

    QariDownloadInfo getSheikhInfoForPosition(int position) {
      return mDownloadInfoMap.get(mQariItems.get(position));
    }

    @Override
    public int getItemCount() {
      return mDownloadInfoMap.size() == 0 ? 0 : mQariItems.size();
    }
  }

  private class SheikhViewHolder extends RecyclerView.ViewHolder {
    public final TextView name;
    public final TextView quantity;
    public final ImageView image;

    SheikhViewHolder(View itemView) {
      super(itemView);
      name = (TextView) itemView.findViewById(R.id.name);
      quantity = (TextView) itemView.findViewById(R.id.quantity);

      image = (ImageView) itemView.findViewById(R.id.image);
      itemView.setOnClickListener(mOnClickListener);
    }
  }
}

File: app/src/main/java/com/quran/labs/androidquran/ui/AudioManagerActivity.kt
package com.quran.labs.androidquran.ui

import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.view.View.OnClickListener
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.ProgressBar
import android.widget.TextView
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.Adapter
import androidx.recyclerview.widget.RecyclerView.ViewHolder
import com.quran.data.core.QuranInfo
import com.quran.labs.androidquran.QuranApplication
import com.quran.labs.androidquran.R
import com.quran.labs.androidquran.common.audio.QariItem
import com.quran.labs.androidquran.util.AudioManagerUtils
import com.quran.labs.androidquran.util.AudioUtils
import com.quran.labs.androidquran.util.QariDownloadInfo
import com.quran.labs.androidquran.util.QuranFileUtils
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import java.util.HashMap
import javax.inject.Inject

class AudioManagerActivity : QuranActionBarActivity() {
  private val disposable: CompositeDisposable = CompositeDisposable()

  private lateinit var progressBar: ProgressBar
  private lateinit var recyclerView: RecyclerView
  private lateinit var shuyookhAdapter: ShuyookhAdapter

  private var basePath: String? = null

  private var qariItems: List<QariItem> = emptyList()

  @Inject
  lateinit var audioUtils: AudioUtils

  @Inject
  lateinit var quranInfo: QuranInfo

  @Inject
  lateinit var quranFileUtils: QuranFileUtils

  override fun onCreate(savedInstanceState: Bundle?) {
    val quranApp = application as QuranApplication
    quranApp.applicationComponent
        .inject(this)
    quranApp.refreshLocale(this, false)

    super.onCreate(savedInstanceState)

    val ab = supportActionBar
    if (ab != null) {
      ab.setTitle(R.string.audio_manager)
      ab.setDisplayHomeAsUpEnabled(true)
    }
    setContentView(R.layout.audio_manager)

    qariItems = audioUtils.getQariList(this)
    shuyookhAdapter = ShuyookhAdapter(qariItems)

    recyclerView = findViewById(R.id.recycler_view)
    recyclerView.setHasFixedSize(true)
    recyclerView.layoutManager = LinearLayoutManager(this)
    recyclerView.itemAnimator = DefaultItemAnimator()
    recyclerView.adapter = shuyookhAdapter

    progressBar = findViewById(R.id.progress)
    basePath = quranFileUtils.getQuranAudioDirectory(this)
  }

  private fun requestShuyookhData() {
    disposable.clear()

    disposable.add(
        AudioManagerUtils.shuyookhDownloadObservable(quranInfo, basePath, qariItems)
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe({ downloadInfo ->
              progressBar.visibility = View.GONE
              shuyookhAdapter.setDownloadInfo(downloadInfo)
              shuyookhAdapter.notifyDataSetChanged()
            }, { })
    )
  }

  override fun onResume() {
    super.onResume()
    requestShuyookhData()
  }

  override fun onDestroy() {
    disposable.dispose()
    super.onDestroy()
  }

  private val onClickListener =
    OnClickListener { v ->
      val position = recyclerView.getChildAdapterPosition(v)
      if (position != RecyclerView.NO_POSITION) {
        val qariItem = shuyookhAdapter.qariItems[position]
        val intent =
          Intent(this@AudioManagerActivity, SheikhAudioManagerActivity::class.java)
        intent.putExtra(SheikhAudioManagerActivity.EXTRA_SHEIKH, qariItem)
        startActivity(intent)
      }
    }

  override fun onOptionsItemSelected(item: MenuItem): Boolean {
    if (item.itemId == android.R.id.home) {
      finish()
      return true
    }
    return super.onOptionsItemSelected(item)
  }

  private inner class ShuyookhAdapter internal constructor(val qariItems: List<QariItem>) :
      Adapter<SheikhViewHolder>() {
    private val inflater: LayoutInflater = LayoutInflater.from(this@AudioManagerActivity)
    private val downloadInfoMap: MutableMap<QariItem, QariDownloadInfo> = HashMap()

    fun setDownloadInfo(downloadInfo: List<QariDownloadInfo>) {
      for (info in downloadInfo) {
        downloadInfoMap[info.qariItem] = info
      }
    }

    override fun onCreateViewHolder(
      parent: ViewGroup,
      viewType: Int
    ): SheikhViewHolder {
      return SheikhViewHolder(inflater.inflate(R.layout.audio_manager_row, parent, false))
    }

    override fun onBindViewHolder(
      holder: SheikhViewHolder,
      position: Int
    ) {
      holder.name.text = qariItems[position].name
      val info = getSheikhInfoForPosition(position)
      val fullyDownloaded = info!!.downloadedSuras.size()
      holder.quantity.text = resources.getQuantityString(
          R.plurals.files_downloaded,
          fullyDownloaded, fullyDownloaded
      )
      if (fullyDownloaded > 0)
        holder.image.setBackgroundResource(R.drawable.downloaded_button_circle)
      else
        holder.image.setBackgroundResource(R.drawable.download_button_circle)
    }

    fun getSheikhInfoForPosition(position: Int): QariDownloadInfo? {
      return downloadInfoMap[qariItems[position]]
    }

    override fun getItemCount(): Int {
      return if (downloadInfoMap.isEmpty()) 0 else qariItems.size
    }
  }

  private inner class SheikhViewHolder internal constructor(itemView: View) :
      ViewHolder(itemView) {
    val name: TextView = itemView.findViewById(R.id.name)
    val quantity: TextView = itemView.findViewById(R.id.quantity)
    val image: ImageView = itemView.findViewById(R.id.image)

    init {
      itemView.setOnClickListener(onClickListener)
    }
  }
}
